rts/motoko-rts/src/stabilization/serialization/stable_memory_stream.rs

```diff
@@ -0,0 +1,149 @@
+//! Streamed read/write access to stable memory.
+//! Supporting Cheney's to-space in stable memory.
+
+use core::mem::{size_of, MaybeUninit};
+
+use crate::{
+    stabilization::{
+        deserialization::stable_memory_access::StableMemoryAccess, grant_stable_space,
+    },
+    stable_mem::{ic0_stable64_read, ic0_stable64_write},
+};
+
+/// Streamed reader/writer on stable memory.
+/// Used for the to-space during stabilization.
+///
+/// The memory supports two location-independent streams:
+/// * Streamed reading and updating, used for scanning and patching pointers.
+/// * Streamed writing, used for allocating new objects.
+pub struct StableMemoryStream {
+    /// The pointers used in the serialized stable memory layout are
+    /// relative to this start address of the to-space.
+    base_address: u64,
+    /// Used for reading and updating.
+    scan_address: u64,
+    /// Used for writing.
+    free_address: u64,
+}
+
+pub trait ScanStream {
+    // Determines whether the stream has reached the end.
+    fn scan_completed(&self) -> bool;
+    // Read a value from the stream.
+    fn read<T>(&mut self) -> T;
+    // Read raw data from the stream.
+    fn raw_read(&mut self, data_address: usize, length: usize);
+    // Go back in the stream.
+    fn rewind(&mut self, length: usize);
```

Comment:
This might be unused?

Reply:
Thanks. Indeed, this is redundant. I have removed it.